mscore/palette.cpp

```diff
@@ -259,9 +259,18 @@ void Palette::contextMenuEvent(QContextMenuEvent* event)
       if (action == clearAction) {
             PaletteCell* cell = cellAt(i);
             if (cell) {
-                  if(cell->tag == "ShowMore")
-                        _moreElements = false;
-                  delete cell;
+                  int ret = QMessageBox::warning(this, QWidget::tr("Delete palette cell"),
+                                                 QWidget::tr("Are you sure you want to delete palette cell %1 ?")
```

Comment:
There should be no variables in brackets. It won't work for translations. Use strings concatenation instead.

Reply:
How do you mean? Like this?

```suggestion
                                                 QWidget::tr("Are you sure you want to delete palette cell ") + %1 + " ?")
```

Reply:
almost, 
`Widget::tr("Are you sure you want to delete palette cell " + cell->name + "?",...`
also drops the excess space before the "?"
But I think we do have other translatable strings with `%N` in them

Reply:
Ok, now I understand. According to the Qt doc (http://doc.qt.io/qt-5/i18n-source-translation.html#using-numbered-arguments), what I had should have worked, though. I think it even did work when I tested it (my language was Afrikaans).

Reply:
It does, use it

Reply:
I am not so sure now that my code is actually translatable. It does not use the tooltip text, somehow. It uses the English name even if the tooltip is in Afrikaans. This does not make sense to me, because in mscore\palette.h, in `struct PaletteCell`, there is this data member:

`QString name;           // used for tool tip`

which is what I used in my dialog's text. The comment after it confuses me, however.

The translatable text is in libmscore\barline.cpp in `const std::vector<BarLineTableItem> BarLine::barLineTable`. I have no idea how that is linked to the palette's tooltip.

If I right-click and choose "Properties...", I see that the name is indeed English, even though its tooltip is Afrikaans. I can't see where the members of `struct PaletteCell` gets initialised, though.

PS: If I change my code as suggested, I get a compile error, so that won't help to try that:

`C:\Users\User\Documents\MuseScore\mscore\palette.cpp:263: error: no matching function for call to 'Ms::Palette::tr(const QString)'
                                                  QWidget::tr("Are you sure you want to delete palette cell " + cell->name + "?"),
                                                                                                                                ^`

Reply:
OK, so you'd need `tr("Are you sure you want to delete palette cell " + QString(cell->name) + "?"),`
But as mentioned, don't do that, use `Widget::tr("Are you sure you want to delete palette cell \"%1\"?",`

Reply:
@L0uisc Take a look at how internationalization engine works in Qt: http://doc.qt.io/qt-5/i18n-source-translation.html. When we build MuseScore, the dictionary is generated. Each tr() string goes there. Then we upload list of all strings to Transifex. Transifex helps people to generate translations and put dictionaries to our storage. After that, dictionaries are available in MuseScore. It means when you use `tr("New string you added yesterday") + cellName` (cellName is an old already translated string) you will get first part in English and cellName in whatever language you use.

@Jojo-Schmitz here is the link of how numbers are processed http://doc.qt.io/qt-5/i18n-source-translation.html#handling-plurals. AFAIU, `tr("Text %2").arg(val1)` means "Text %2" is a string which goes directly to transifex. So, translators shoud keep %2 there to keep it working. This is ugly. You can use several strings instead. So, translators will just translate first part "Are you sure you want to delete palette cell". @L0uisc To make the code work, you need to convert raw strings to QStrings.

Reply:
It is used in MusaeScore since ages and in numorous places, check https://www.transifex.com/musescore/musescore/translate/#de/musescore/41206011?q=text%3A%251

Reply:
We did at a time (well, I did) move HTML code out of the translations, but this isn't the case here

Reply:
But you do have a point here, cellName might be translated already.

Reply:
still %1 would work

Reply:
still that stray space between the %1 and the ? is too much...

Reply:
> @L0uisc Take a look at how internationalization engine works in Qt: http://doc.qt.io/qt-5/i18n-source-translation.html. When we build MuseScore, the dictionary is generated. Each tr() string goes there. Then we upload list of all strings to Transifex. Transifex helps people to generate translations and put dictionaries to our storage. After that, dictionaries are available in MuseScore. It means when you use `tr("New string you added yesterday") + cellName` (cellName is an old already translated string) you will get first part in English and cellName in whatever language you use.

@anatoly-os As I understand it, the `%1` argument can also be set to be an already translated string. The problem I have with doing things like you suggested, is that it fixes the word order to the correct English, but not necessarily for other languages. In Afrikaans, for example, I would like to translate `QWidget::tr("Are you sure you want to delete palette cell \"%1\"?"` to something like `"Is jy seker jy wil paletsel \"%1\" uitvee?"` The word after the `%1` is the verb "delete". Separating and concatenating strings does not allow that. Also, see my link to the Qt doc in [this](https://github.com/musescore/MuseScore/pull/4141#discussion_r233566726) comment above.

Reply:
Ah, yes, thanks for reminding me of this very good reson to use these `%N` stuff ;-)

Reply:
How about this: http://doc.qt.io/qt-5/i18n-source-translation.html#translator-comments

Reply:
I don't know if this will work, but as I understand. this should add the string after the `//: ` to the Context field. I also cleaned up the logic after the messagebox to make it more readable.

`int ret = QMessageBox::warning(this, QWidget::tr("Delete palette cell"),
      //: %1 is a variable for the cell's name. Your translation should contain it.
     QWidget::tr("Are you sure you want to delete palette cell \"%1\"?")
     .arg(cell->name), QMessageBox::Yes | QMessageBox::No,
     QMessageBox::Yes);
     if (ret != QMessageBox::Yes)
           return;
     if(cell->tag == "ShowMore")
           _moreElements = false;
     delete cell;`

Reply:
As mentioned earlied this `%N` mechanism is used in the translations since ages and in many places, translators know about it, so no need to 'decorate' every occurence with sich a comment

Reply:
Ok. @anatoly-os I will start a new PR to do the following:

- [ ] Remove the redundant space after the `%1` variable

- [ ] Put the `%1` variable in quotes

- [ ] Simplify the logic after the messagebox

Is this the right way (starting a new PR), or should I do something else to make these changes?

Reply:
Starting the new one is ok

Reply:
BTW: another reason for using `%N` rather concatenating strings are the translation for RTL languages, Arabic, Hebrew, etc.